Engine/platform/windows/gfx/ali3dd3d.cpp

```diff
@@ -1335,10 +1335,53 @@ void D3DGraphicsDriver::_render(bool clearDrawListAfterwards)
     _filter->SetSamplerStateForStandardSprite(direct3ddevice);
     D3DTEXTUREFILTERTYPE filterType;
     direct3ddevice->GetSamplerState(0, D3DSAMP_MAGFILTER, (DWORD*)&filterType);
-    if (direct3ddevice->StretchRect(pNativeSurface, NULL, pBackBuffer, &viewport_rect, filterType) != D3D_OK)
+    
+    const int pixelSize = viewport_rect.right / _srcRect.GetWidth();
+
+    if (pixelSize == 1) {
+        // since the pixel size is 1 the distortion will not happen
+        if (direct3ddevice->StretchRect(pNativeSurface, NULL, pBackBuffer, &viewport_rect, filterType) != D3D_OK)
+        {
+            throw Ali3DException("IDirect3DSurface9::StretchRect failed");
+        }
+    } 
+    else // TODO find a way to make this workaround unnecessary
     {
-      throw Ali3DException("IDirect3DSurface9::StretchRect failed");
+        // Bug: StretchRect makes the last right and bottom pixel one single pixel, always, no matter how big you stretch (reason uknown)
```

Comment:
I wonder if we're working around a driver issue here. I read here ( http://sjbrown.co.uk/2003/05/01/fix-directx-rasterisation/ ) that some drivers do not take into account half texels for StretchRect, so maybe this is a similar issue?

Reply:
I experimented with using the direct3d reference device, and you do indeed get doubled up rows with this patch:
![image](https://user-images.githubusercontent.com/335451/67547928-21505a00-f74c-11e9-85fd-9a29f86a5138.png)

Reply:
I don't know, many other places seem to use shaders to perform the scaling or textures drawn with quads.
There are no mentions of this issue where StretchRect is used, so I wonder is it just us or is it overlooked?
I think I'd rather implement a D3D10 driver rather than trying again to fix something I haven't been able to fix since the first time we found out about the issue.

Reply:
I think drawing with a textured quad would certainly give us more control over the problem. If you don't mind waiting, this might be better suited for after the SDL port goes in, as we rely on SDL to do the window management.